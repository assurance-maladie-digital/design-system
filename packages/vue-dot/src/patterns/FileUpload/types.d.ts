import { ErrorCodes } from './errorCodes';

export interface ErrorEvent {
	file: File | DataTransferItemList;
	code: ErrorCodes;
}

export type ErrorCodesType = {
	[key: string]: string;
};

export interface HTMLInputEvent extends Event {
	target: HTMLInputElement & EventTarget;
}
