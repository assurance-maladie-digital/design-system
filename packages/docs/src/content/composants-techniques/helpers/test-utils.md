---
title: testUtils
description: Le helper `testUtils` contient un ensemble de fonctions utilitaires pour les tests unitaires.
---

<doc-tabs>

<doc-tab-item label="Utilisation">

<doc-alert type="info">

Le helper `testUtils` exporte également toutes les fonctions du package [Vue Test Utils](https://vue-test-utils.vuejs.org/).

</doc-alert>

### applyGlobalConfig

La fonction `applyGlobalConfig` permet d’appliquer la configuration globale pour les tests, elle est utilisée dans le fichier `setup.ts`.

```ts
import { applyGlobalConfig } from '@cnamts/vue-dot/src/helpers/testUtils';

applyGlobalConfig();
```

### installGlobalPlugins

La fonction `installGlobalPlugins` permet d’installer les plugins [Vue Input Facade](https://ronaldjerez.github.io/vue-input-facade/), [Vue Meta](https://vue-meta.nuxtjs.org/), [Vuex](https://vuex.vuejs.org/), [Vuetify](https://v2.vuetifyjs.com/) et Vue Dot pour les tests.

```ts
import { createLocalVue, installGlobalPlugins } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installGlobalPlugins(localVue);
```

### installRouter

La fonction `installRouter` permet d’installer [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createLocalVue, installRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installRouter(localVue);
```

### createRouter

La fonction `createRouter` permet de créer une instance de [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const routes = [];

const router = createRouter({
	routes
});
```

### createStore

La fonction `createStore` permet de créer une instance de [Vuex](https://vuex.vuejs.org/) pour les tests.

```ts
import { createStore } from '@cnamts/vue-dot/src/helpers/testUtils';

const store = createStore({
	state: {}
});
```

### createVuetifyInstance

La fonction `createVuetifyInstance` permet de créer une instance de [Vuetify](https://v2.vuetifyjs.com/) pour les tests.

```ts
import { createVuetifyInstance } from '@cnamts/vue-dot/src/helpers/testUtils';

const vuetify = createVuetifyInstance();
```

### mockVFormRef

La fonction `mockVFormRef` permet de mocker la référence d’un composant [VForm](https://v2.vuetifyjs.com/en/components/forms/) pour les tests.

```ts
import { mockVFormRef } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	$refs: {
		form: mockVFormRef(true)
	}
};
```

### mockTranslations

La fonction `mockTranslations` permet de mocker les traductions pour les tests.

```ts
import { mockTranslations } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	...mockTranslations<string | string[]>({
		'path.to.title': 'Titre',
		'path.to.items': [
			'Élément 1',
			'Élément 2',
			'Élément 3'
		]
	})
};
```

</doc-tab-item>

<doc-tab-item label="API">
<doc-api name="helpers/test-utils"></doc-api>
</doc-tab-item>

</doc-tabs>
