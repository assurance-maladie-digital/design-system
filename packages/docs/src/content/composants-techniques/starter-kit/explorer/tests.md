Le dossier `tests` contient les tests d’interface et les tests unitaires globaux de l’application.

<doc-alert type="warning">
Il n’est pas recommandé de renommer ce dossier.
</doc-alert>
