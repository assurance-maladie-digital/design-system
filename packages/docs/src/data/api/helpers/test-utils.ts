import { Api } from '~/types';

export const api: Api = {
	applyGlobalConfig: {
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	installGlobalPlugins: {
		arguments: [
			{
				name: 'localVue',
				description: 'Local Vue instance',
				type: 'Vue',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	installRouter: {
		arguments: [
			{
				name: 'localVue',
				description: 'Local Vue instance',
				type: 'Vue',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	createRouter: {
		arguments: [
			{
				name: 'options',
				description: 'Router options',
				type: 'RouterOptions',
				required: false,
			},
		],
		returnValue: [
			{
				type: 'VueRouter',
			},
		],
	},
	createStore: {
		arguments: [
			{
				name: 'options',
				description: 'Store options',
				type: 'StoreOptions',
				required: false,
			},
		],
		returnValue: [
			{
				type: 'Store',
			},
		],
	},
	createVuetifyInstance: {
		arguments: [
			{
				name: 'preset',
				description: 'Vuetify preset',
				type: 'Partial<UserVuetifyPreset>',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'Vuetify',
			},
		],
	},
	mockVFormRef: {
		arguments: [
			{
				name: 'isValid',
				description: 'Is valid',
				type: 'boolean',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	mockTranslations: {
		arguments: [
			{
				name: 'translations',
				description: 'Translations',
				type: 'IndexedObject',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
};
