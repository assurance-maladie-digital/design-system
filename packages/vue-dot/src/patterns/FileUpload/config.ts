export const config = {
	icon: {
		size: 40,
		color: 'primary'
	}
};
