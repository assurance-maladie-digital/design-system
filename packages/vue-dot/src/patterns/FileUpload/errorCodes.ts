export enum ErrorCodes {
	MULTIPLE_FILES_SELECTED = 'MULTIPLE_FILES_SELECTED',
	FILE_TOO_LARGE = 'FILE_TOO_LARGE',
	FILE_EXTENSION_NOT_ALLOWED = 'FILE_EXTENSION_NOT_ALLOWED'
}
