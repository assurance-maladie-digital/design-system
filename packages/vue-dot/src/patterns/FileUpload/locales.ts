export const locales = {
	or: 'Ou',
	chooseFile: 'Choisir un fichier',
	infoText: (max: string, ext: string, length: number): string => `Taille max. : ${max}. ${length === 1 ? 'Format accepté' : 'Formats acceptés'} : ${ext}`,
	fileSizeUnits: [
		'o',
		'Ko',
		'Mo',
		'Go',
		'To'
	]
};
