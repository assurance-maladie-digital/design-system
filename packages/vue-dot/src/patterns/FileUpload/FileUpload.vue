<template>
	<label
		v-ripple="!noRipple"
		:style="widthStyles"
		:class="[
			{
				'dragover': dragover,
				'dark-mode': $vuetify.theme.dark
			},
			colors.label
		]"
		class="vd-file-upload d-block pa-4"
		@dragover.prevent="dragover = true"
		@dragleave="dragover = false"
		@drop.prevent="dropHandler"
	>
		<input
			ref="vdInputEl"
			type="file"
			:disabled="disabled"
			:multiple="multiple"
			:accept="computedAccept"
			class="vd-file-upload-input"
			@change="inputValueChanged"
		>

		<slot name="placeholder">
			<span class="vd-file-upload-placeholder">
				<slot name="icon">
					<VIcon v-bind="options.icon">
						{{ uploadIcon }}
					</VIcon>
				</slot>

				<span
					:class="colors.multiple"
					class="mt-1 font-weight-medium"
				>
					<slot
						name="action-text"
						:multiple="multiple"
					>
						{{ multiple ? 'Placez vos fichiers ici' : 'Placez votre fichier ici' }}
					</slot>
				</span>

				<span
					:class="colors.info"
					class="mb-2"
				>
					<slot name="or">
						{{ locales.or }}
					</slot>
				</span>

				<span class="vd-file-upload-btn primary white--text text-uppercase py-2 px-4 elevation-2">
					<slot name="button-text">
						{{ locales.chooseFile }}
					</slot>
				</span>

				<span
					:class="colors.info"
					class="mt-4 text-body-2 font-weight-regular"
				>
					<slot
						name="info-text"
						:max-size="maxSizeReadable"
						:extensions="extensions"
					>
						{{ locales.infoText(maxSizeReadable, extensions, allowedExtensions.length) }}
					</slot>
				</span>
			</span>
		</slot>
	</label>
</template>

<script lang="ts">
	import Vue, { PropType } from 'vue';
	import Component, { mixins } from 'vue-class-component';

	import { config } from './config';
	import { locales } from './locales';

	import { FileUploadCore } from './mixins/fileUploadCore';
	import { Widthable } from '../../mixins/widthable';

	import { customizable } from '../../mixins/customizable';

	import { calcHumanFileSize } from '../../functions/calcHumanFileSize';

	import { mdiCloudUpload } from '@mdi/js';

	import { IndexedObject } from '../../types';

	const Props = Vue.extend({
		props: {
			value: {
				// File is not a valid prop type,
				// use null to allow any type & provide custom validation
				type: null as unknown as PropType<File | File[] | null>,
				default: null,
				/** @see https://github.com/vuetifyjs/vuetify/blob/master/packages/vuetify/src/components/VFileInput/VFileInput.ts#L71 */
				validator(val): boolean {
					if (val === null) {
						return true;
					}

					const value = Array.isArray(val) ? val : [val];
					const isValid = value.every((v) => v !== null && typeof v === 'object');

					return isValid;
				}
			},
			multiple: {
				type: Boolean,
				default: false
			},
			noRipple: {
				type: Boolean,
				default: false
			},
			disabled: {
				type: Boolean,
				default: false
			}
		}
	});

	const MixinsDeclaration = mixins(Props, customizable(config), FileUploadCore, Widthable);

	@Component({
		model: {
			prop: 'value',
			event: 'change'
		}
	})
	export default class FileUpload extends MixinsDeclaration {
		locales = locales;

		uploadIcon = mdiCloudUpload;

		get colors(): IndexedObject {
			return {
				label: 'primary--text',
				multiple: 'black--text',
				info: 'grey--text'
			};
		}

		get maxSizeReadable(): string {
			return calcHumanFileSize(this.fileSizeMax, this.fileSizeUnits);
		}

		get extensions(): string {
			return this.allowedExtensions.join(', ').toUpperCase();
		}

		public retry(): void {
			this.$refs.vdInputEl.click();
		}
	}
</script>

<style lang="scss" scoped>
	@import '@cnamts/design-tokens/dist/tokens';

	.vd-file-upload {
		cursor: pointer;
		position: relative;
		border: 1px dashed;
		background: #fff;
		transition: background .25s;

		&:hover,
		&:focus-within,
		&.dragover {
			background: $vd-parme-lighten-90;
		}

		&.dark-mode {
			&:hover,
			&:focus-within,
			&.dragover {
				background: #303030;
			}
		}
	}

	.vd-file-upload-placeholder {
		display: flex;
		align-items: center;
		flex-direction: column;
	}

	.vd-file-upload-input {
		position: absolute;
		width: 1px;
		height: 1px;
		white-space: nowrap;
		overflow: hidden;
		clip: rect(1px, 1px, 1px, 1px);
	}

	.vd-file-upload-btn {
		border-radius: 4px;
		transition: background .25s;
	}
</style>
