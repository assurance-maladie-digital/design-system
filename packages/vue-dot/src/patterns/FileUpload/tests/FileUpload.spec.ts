import Vue from 'vue';
import { Wrapper } from '@vue/test-utils';

import { mountComponent } from '@/tests';

import FileUpload from '../';

let wrapper: Wrapper<Vue>;

describe('FileUpload', () => {
	it('renders correctly', () => {
		wrapper = mountComponent(FileUpload);

		expect(wrapper).toMatchSnapshot();
	});

	it('renders correctly with only one extension allowed', () => {
		wrapper = mountComponent(FileUpload, {
			propsData: {
				allowedExtensions: ['pdf']
			}
		});

		expect(wrapper).toMatchSnapshot();
	});
});
