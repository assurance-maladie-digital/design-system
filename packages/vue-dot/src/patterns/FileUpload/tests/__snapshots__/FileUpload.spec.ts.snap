// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileUpload renders correctly 1`] = `
<label class="vd-file-upload d-block pa-4 primary--text" style="width: 100%;"><input type="file" accept=".pdf,.jpg,.jpeg,.png" class="vd-file-upload-input"> <span class="vd-file-upload-placeholder"><vicon-stub size="40" color="primary">
					M11 20H6.5Q4.22 20 2.61 18.43 1 16.85 1 14.58 1 12.63 2.17 11.1 3.35 9.57 5.25 9.15 5.88 6.85 7.75 5.43 9.63 4 12 4 14.93 4 16.96 6.04 19 8.07 19 11 20.73 11.2 21.86 12.5 23 13.78 23 15.5 23 17.38 21.69 18.69 20.38 20 18.5 20H13V12.85L14.6 14.4L16 13L12 9L8 13L9.4 14.4L11 12.85Z
				</vicon-stub> <span class="mt-1 font-weight-medium black--text">
					Placez votre fichier ici
				</span> <span class="mb-2 grey--text">
					Ou
				</span> <span class="vd-file-upload-btn primary white--text text-uppercase py-2 px-4 elevation-2">
					Choisir un fichier
				</span> <span class="mt-4 text-body-2 font-weight-regular grey--text">
					Taille max. : 10 Mo. Formats acceptés : PDF, JPG, JPEG, PNG
				</span></span></label>
`;

exports[`FileUpload renders correctly with only one extension allowed 1`] = `
<label class="vd-file-upload d-block pa-4 primary--text" style="width: 100%;"><input type="file" accept=".pdf" class="vd-file-upload-input"> <span class="vd-file-upload-placeholder"><vicon-stub size="40" color="primary">
					M11 20H6.5Q4.22 20 2.61 18.43 1 16.85 1 14.58 1 12.63 2.17 11.1 3.35 9.57 5.25 9.15 5.88 6.85 7.75 5.43 9.63 4 12 4 14.93 4 16.96 6.04 19 8.07 19 11 20.73 11.2 21.86 12.5 23 13.78 23 15.5 23 17.38 21.69 18.69 20.38 20 18.5 20H13V12.85L14.6 14.4L16 13L12 9L8 13L9.4 14.4L11 12.85Z
				</vicon-stub> <span class="mt-1 font-weight-medium black--text">
					Placez votre fichier ici
				</span> <span class="mb-2 grey--text">
					Ou
				</span> <span class="vd-file-upload-btn primary white--text text-uppercase py-2 px-4 elevation-2">
					Choisir un fichier
				</span> <span class="mt-4 text-body-2 font-weight-regular grey--text">
					Taille max. : 10 Mo. Format accepté : PDF
				</span></span></label>
`;
